tool-openssl/pkcs8.cc

```diff
@@ -0,0 +1,262 @@
+      fprintf(stderr, "Error: failed to determine input file size\n");
+      return false;
+    }
+    
+    data = (uint8_t *)OPENSSL_malloc(len);
+    if (!data || fread(data, 1, len, in_file.get()) != (size_t)len) {
+      OPENSSL_free(data);
+      OPENSSL_free(passin);
+      OPENSSL_free(passout);
+      fprintf(stderr, "Error: failed to read input file\n");
+      return false;
+    }
+    
+    const uint8_t *p = data;
+    pkey = d2i_AutoPrivateKey(nullptr, &p, len);  // Support for all key types
+    OPENSSL_free(data);
+  }
+  
+  if (!pkey) {
+    OPENSSL_free(passin);
+    OPENSSL_free(passout);
+    fprintf(stderr, "Error: Failed to read private key from '%s'\n", in_path.c_str());
+    fprintf(stderr, "Check that the file contains a valid key and the password (if any) is correct\n");
+    ERR_print_errors_fp(stderr);
+    return false;
+  }
+
+  // Process according to options
+  bool result = false;
+  if (topk8) {
+    // Convert to PKCS#8
+    if (nocrypt) {
+      // Unencrypted PKCS#8
+        PKCS8_PRIV_KEY_INFO *p8inf = EVP_PKEY2PKCS8(pkey);
+        if (!p8inf) {
+          fprintf(stderr, "Error: Failed to convert key to PKCS#8 format\n");
+          fprintf(stderr, "The key type may not be supported for PKCS#8 conversion\n");
+          ERR_print_errors_fp(stderr);
+      } else {
+        // Write the output
+        if (outform == "PEM") {
+          result = PEM_write_PKCS8_PRIV_KEY_INFO(out, p8inf);
+        } else {  // DER
+          result = i2d_PKCS8_PRIV_KEY_INFO_fp(out, p8inf);
+        }
+        PKCS8_PRIV_KEY_INFO_free(p8inf);
+      }
+    } else {
+      // Encrypted PKCS#8
+      if (!v2_cipher.empty()) {
+        // Use PKCS#5 v2.0
+        const EVP_CIPHER *cipher = EVP_get_cipherbyname(v2_cipher.c_str());
+        if (!cipher) {
+          fprintf(stderr, "Error: Unknown cipher %s\n", v2_cipher.c_str());
+          EVP_PKEY_free(pkey);
+          OPENSSL_free(passin);
+          OPENSSL_free(passout);
+          return false;
+        }
+        
+        // Handle PRF algorithm if specified
+        // Note: While OpenSSL supports explicit PRF selection, AWS-LC uses its default PRF
+        // implementation for PKCS#8 encryption. The -v2prf parameter is accepted for
+        // command-line compatibility but does not affect the actual PRF used.
+        if (!v2prf.empty()) {
+          // In AWS-LC, we'll validate that it's one of the known values to maintain compatibility
+          if (v2prf != "hmacWithSHA1" && 
+              v2prf != "hmacWithSHA256" && 
+              v2prf != "hmacWithSHA512") {
+            fprintf(stderr, "Error: Unknown PRF %s\n", v2prf.c_str());
+            EVP_PKEY_free(pkey);
+            OPENSSL_free(passin);
+            OPENSSL_free(passout);
+            return false;
+          }
+          // The PRF specification is validated but not used by AWS-LC implementation
+        }
+        
+        // Convert and encrypt
+        X509_SIG *p8 = nullptr;
```

Comment:
Smart pointer here as well. Ideally we wouldn't need any `*_free` calls in C++ code.